// Compile _pb2.py with 
// protoc -I=$SRC_DIR --python_out=$DST_DIR $SRC_DIR/cc1_levelset.proto
// or for same folder ops just
// protoc --python_out=. cc1_levelset.proto 

syntax = "proto3";

package cc1_levelset_utils;

message Levelset {
    string name = 1;
    repeated Level levels = 2;
    repeated Story stories = 3;


    message Story {
        int32 level_number = 1;
        Type type = 2;
        repeated string text = 3;

        enum Type {
            PROLOGUE = 0;
            EPILOGUE = 1;
        }
    }
}

message Level {
    string title = 1;
    string author = 2;
    int32 number = 3;
    int32 time = 4;
    int32 chips = 5;
    string hint = 6;
    string password = 7;

    // All int32 values must be packed and unpacked by these formulas:
    //          p = y * 32 + x
    //       x, y = p % 32, p // 32
    map<int32, TileSpec> map = 8; 
    map<int32, int32> trap_controls = 9;
    map<int32, int32> clone_controls = 10;
    repeated int32 movement = 11;
}

message TileSpec {
    CC1TileCode top = 1;
    CC1TileCode bottom = 2;
}

enum CC1TileCode {
    FLOOR = 0;
    WALL = 1;
    CHIP = 2;
    WATER = 3;
    FIRE = 4;
    INV_WALL_PERM = 5;
    PANEL_N = 6;
    PANEL_W = 7;
    PANEL_S = 8;
    PANEL_E = 9;
    BLOCK = 10;
    DIRT = 11;
    ICE = 12;
    FORCE_S = 13;
    CLONE_BLOCK_N = 14;
    CLONE_BLOCK_W = 15;
    CLONE_BLOCK_S = 16;
    CLONE_BLOCK_E = 17;
    FORCE_N = 18;
    FORCE_E = 19;
    FORCE_W = 20;
    EXIT = 21;
    BLUE_DOOR = 22;
    RED_DOOR = 23;
    GREEN_DOOR = 24;
    YELLOW_DOOR = 25;
    ICE_SE = 26;
    ICE_SW = 27;
    ICE_NW = 28;
    ICE_NE = 29;
    BLUE_WALL_FAKE = 30;
    BLUE_WALL_REAL = 31;
    NOT_USED_0 = 32;
    THIEF = 33;
    SOCKET = 34;
    GREEN_BUTTON = 35;
    CLONE_BUTTON = 36;
    TOGGLE_WALL = 37;
    TOGGLE_FLOOR = 38;
    TRAP_BUTTON = 39;
    TANK_BUTTON = 40;
    TELEPORT = 41;
    BOMB = 42;
    TRAP = 43;
    INV_WALL_APP = 44;
    GRAVEL = 45;
    POP_UP_WALL = 46;
    HINT = 47;
    PANEL_SE = 48;
    CLONER = 49;
    FORCE_RANDOM = 50;
    DROWN_CHIP = 51;
    BURNED_CHIP0 = 52;
    BURNED_CHIP1 = 53;
    NOT_USED_1 = 54;
    NOT_USED_2 = 55;
    NOT_USED_3 = 56;
    CHIP_EXIT = 57;
    UNUSED_EXIT_0 = 58;
    UNUSED_EXIT_1 = 59;
    CHIP_SWIMMING_N = 60;
    CHIP_SWIMMING_W = 61;
    CHIP_SWIMMING_S = 62;
    CHIP_SWIMMING_E = 63;
    ANT_N = 64;
    ANT_W = 65;
    ANT_S = 66;
    ANT_E = 67;
    FIREBALL_N = 68;
    FIREBALL_W = 69;
    FIREBALL_S = 70;
    FIREBALL_E = 71;
    BALL_N = 72;
    BALL_W = 73;
    BALL_S = 74;
    BALL_E = 75;
    TANK_N = 76;
    TANK_W = 77;
    TANK_S = 78;
    TANK_E = 79;
    GLIDER_N = 80;
    GLIDER_W = 81;
    GLIDER_S = 82;
    GLIDER_E = 83;
    TEETH_N = 84;
    TEETH_W = 85;
    TEETH_S = 86;
    TEETH_E = 87;
    WALKER_N = 88;
    WALKER_W = 89;
    WALKER_S = 90;
    WALKER_E = 91;
    BLOB_N = 92;
    BLOB_W = 93;
    BLOB_S = 94;
    BLOB_E = 95;
    PARAMECIUM_N = 96;
    PARAMECIUM_W = 97;
    PARAMECIUM_S = 98;
    PARAMECIUM_E = 99;
    BLUE_KEY = 100;
    RED_KEY = 101;
    GREEN_KEY = 102;
    YELLOW_KEY = 103;
    FLIPPERS = 104;
    FIRE_BOOTS = 105;
    SKATES = 106;
    SUCTION_BOOTS = 107;
    PLAYER_N = 108;
    PLAYER_W = 109;
    PLAYER_S = 110;
    PLAYER_E = 111;  
}